zcash_client_backend/src/tor.rs

```diff
@@ -0,0 +1,104 @@
+//! Tor support for Zcash wallets.
+
+use std::{fmt, io, path::Path};
+
+use arti_client::{config::TorClientConfigBuilder, TorClient};
+use tor_rtcompat::PreferredRuntime;
+use tracing::debug;
+
+mod http;
+
+/// A Tor client that exposes capabilities designed for Zcash wallets.
+pub struct Client {
+    inner: TorClient<PreferredRuntime>,
+}
+
+impl Client {
+    /// Creates and bootstraps a Tor client.
+    ///
+    /// The client's persistent data and cache are both stored in the given directory.
+    /// Preserving the contents of this directory will speed up subsequent calls to
+    /// `Client::create`.
+    ///
+    /// Returns an error if `tor_dir` does not exist, or if bootstrapping fails.
+    pub async fn create(tor_dir: &Path) -> Result<Self, Error> {
+        let runtime = PreferredRuntime::current()?;
+
+        if !tokio::fs::try_exists(tor_dir).await? {
+            return Err(Error::MissingTorDirectory);
+        }
+
+        let mut config_builder = TorClientConfigBuilder::from_directories(
+            tor_dir.join("arti-data"),
+            tor_dir.join("arti-cache"),
+        );
+        config_builder
+            .storage()
+            .permissions()
+            .dangerously_trust_everyone();
+        let config = config_builder
+            .build()
+            .expect("all required fields initialized");
+
+        let client_builder = TorClient::with_runtime(runtime).config(config);
+
+        debug!("Bootstrapping Tor");
+        let inner = client_builder.create_bootstrapped().await?;
+        debug!("Tor bootstrapped");
+
+        Ok(Self { inner })
+    }
+}
+
+/// Errors that can occur while creating or using a Tor [`Client`].
+#[derive(Debug)]
+pub enum Error {
+    /// The directory passed to [`Client::create`] does not exist.
+    MissingTorDirectory,
```

Comment:
Does it make sense to return the path?

Reply:
I don't think so? The caller directly provided the path as the only argument, and therefore has it. If they want to augment the error response with that path as context, they can do so.

The other direction we could go here is to just create the folder ourselves, but I feel like it should be a conscious choice of the caller to use a particular directory, especially given that there's no difference between creating a new Arti client, and loading an existing one.